<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
 <html><head><title>R: Try an Expression Allowing Error Recovery</title>
 <meta http-equiv="Content-Type" content="text/html; charset=utf-8">
 <link rel="stylesheet" type="text/css" href="R.css">
 </head><body>
 
 <table width="100%" summary="page for try"><tr><td>try</td><td align="right">R Documentation</td></tr></table>
 
 <h2>Try an Expression Allowing Error Recovery</h2>
 
 <h3>Description</h3>
 
 
 <p><code>try</code> is a wrapper to run an expression that might fail and allow
 the user's code to handle error-recovery.
 </p>
 
 
 <h3>Usage</h3>
 
 <pre>
 try(expr, silent = FALSE)
 </pre>
 
 
 <h3>Arguments</h3>
 
 
 <table summary="R argblock">
 <tr valign="top"><td><code>expr</code></td>
 <td>
 <p>an <font face="Courier New,Courier" color="#666666"><b>R</b></font> expression to try.</p>
 </td></tr>
 <tr valign="top"><td><code>silent</code></td>
 <td>
 <p>logical: should the report of error messages be suppressed?</p>
 </td></tr>
 </table>
 
 
 <h3>Details</h3>
 
 
 <p><code>try</code> evaluates an expression and traps any errors that occur
 during the evaluation.  If an error occurs then the error
 message is printed to the <code>stderr</code> connection unless
 <code>options("show.error.messages")</code> is false or
 the call includes <code>silent = TRUE</code>.  The error message is also
 stored in a buffer where it can be retrieved by
 <code>geterrmessage</code>. (This should not be needed as the value returned
 in case of an error contains the error message.)
 </p>
 <p><code>try</code> is implemented using <code>tryCatch</code>; for
 programming, instead of <code>try(expr, silent=TRUE)</code>, something like
 <code>tryCatch(expr, error = function(e) e)</code> (or other simple
 error handler functions) may be more efficient and flexible.
 </p>
 
 
 <h3>Value</h3>
 
 
 <p>The value of the expression if <code>expr</code> is evaluated without error,
 but an invisible object of class <code>"try-error"</code> containing the
 error message, and the error condition as the <code>"condition"</code>
 attribute, if it fails.
 </p>
 
 
 <h3>See Also</h3>
 
 
 <p><code>options</code> for setting error handlers and suppressing the
 printing of error messages;
 <code>geterrmessage</code> for retrieving the last error message.
 <code>tryCatch</code> provides another means of catching and handling
 errors.
 </p>
 
 
 <h3>Examples</h3>
 
 <pre>
 ## this example will not work correctly in example(try), but
 ## it does work correctly if pasted in
 options(show.error.messages = FALSE)
 try(log("a"))
 print(.Last.value)
 options(show.error.messages = TRUE)
 
 ## alternatively,
 print(try(log("a"), TRUE))
 
 ## run a simulation, keep only the results that worked.
 set.seed(123)
 x &lt;- stats::rnorm(50)
 doit &lt;- function(x)
 {
     x &lt;- sample(x, replace=TRUE)
     if(length(unique(x)) &gt; 30) mean(x)
     else stop("too few unique points")
 }
 ## alternative 1
 res &lt;- lapply(1:100, function(i) try(doit(x), TRUE))
 ## alternative 2
 ## Not run: res &lt;- vector("list", 100)
 for(i in 1:100) res[[i]] &lt;- try(doit(x), TRUE)
 ## End(Not run)
 unlist(res[sapply(res, function(x) !inherits(x, "try-error"))])
 </pre>
 
 
 </body></html>
